glean-core/src/upload/mod.rs

```diff
@@ -0,0 +1,160 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Manages de pending pings queue and directory.
+
+// !IMPORTANT!
+// Remove the next line when this module's
+// functionality is in the Glean object. This
+// is here just to not have lint error for now.
+#![allow(dead_code)]
+
+use std::collections::VecDeque;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::sync::{Arc, RwLock};
+use std::thread;
+
+use log;
+use serde_json::Value as JsonValue;
+
+use request::PingRequest;
+use util::process_pings_dir;
+
+mod request;
+mod util;
+
+/// Manages the pending pings queue and directory.
+#[derive(Debug)]
+pub struct PingUploadHelper {
+    /// A FIFO queue storing a `PingRequest` for each pending_ping.
+    queue: Arc<RwLock<VecDeque<PingRequest>>>,
+}
+
+impl PingUploadHelper {
+    /// Create a new PingUploadHelper.
+    ///
+    /// This will spawn a new thread and processes the pending pings
+    /// folder filling up the queue with whatever pings are in there
+    /// ordered by oldest to newest file created.
+    pub fn new(data_path: &str) -> Self {
+        let queue = Arc::new(RwLock::new(VecDeque::new()));
+
+        let data_path = PathBuf::from_str(&data_path).expect("data_path must be a valid path.");
+        let local_queue = queue.clone();
+        thread::spawn(move || match process_pings_dir(&data_path) {
+            Ok(requests) => {
+                let mut local_queue = local_queue
+                    .write()
+                    .expect("Can't write to pending pings queue.");
+                for request in requests {
+                    local_queue.push_back(request);
+                }
+            }
+            Err(e) => log::info!("Error processing pending pings directories! {}", e),
+        })
+        .join()
+        .expect("Could not process pending pings.");
+
+        Self { queue }
+    }
+
+    /// Creates a `PingRequest` and adds it to the queue.
+    pub fn enqueue_ping(&self, uuid: &str, url: &str, body: JsonValue) {
+        let mut queue = self
+            .queue
+            .write()
```

Comment:
At that point do we _really_ care about the exact order?
They are already delayed if they are read on startup, they will get processed eventually.

Reply:
> 
> 
> At that point do we _really_ care about the exact order?

That's a good point. One thing we could do is leave it as it is and document this caveat. We can get to that in the future, if needed.

Reply:
> To understand how to structure this a bit better, I'd also recommend taking a look at how Firefox Desktop handles this :-) See, for example, this code (and follow along there)

From what I see in this code, desktop works with a flag signaling if processing of pending pings on disk is done. That is what I have started to do with 0f7dc5b and what is described in https://github.com/mozilla/glean/pull/673#discussion_r375815726.

For the sake of not making this PR any longer than it already is, I think it would be best to finish it now and have follow ups for:

- Using the flag to return different statuses inclusing the "come back later" status from `get_next_ping`;
- Adding support for deletion request pings;
- Adding functionatility to process upload responses;
- Adding all of it to the Glean object;
- And then finally adding it to the FFI.

As suggested by @badboy on Riot, we can put all this behind a feature flag for now, if we think necessary. It probably is necessary because even though this code doesn't interefere with the way ping uploading is done on the platform side, it does add tasks to init which is not ideal if it's not going to be used anywhere.

Let me know your thoughts :)

Reply:
Works for me!

Reply:
Perfect, I'll file bugs for all the things listed there.